File: Sorting Algorithims/MergeSort/mergesort.py
def merge_sort(arr):
    # Divide Array Into 2 Subpieces
    if len(arr) > 1:
        left_arr = arr[:len(arr)//2]
        right_arr = arr[len(arr)//2:]


        # Recursion
        merge_sort(left_arr)
        merge_sort(right_arr)

        # Merge
        i = 0 # left idx
        j = 0 # right idx
        k = 0 # merge idx

        while i < len(left_arr) and j < len(right_arr):
            if left_arr[i] < right_arr[j]:
                arr[k] = left_arr[i]
                i += 1
            else:
                arr[k] = right_arr[j]
                j += 1
            k += 1

        while i < len(left_arr):
            arr[k] = left_arr[i]
            i += 1
            k += 1

        while j < len(right_arr):
            arr[k] = right_arr[j]
            j += 1
            k += 1
    return arr

arr_test = [2, 3, 5, 1, 7, 4, 4, 4, 2, 6, 0]

print(merge_sort(arr_test))

File: Sorting Algorithims/MergeSort/mergesort.java
// Java program for implementation of Merge Sort
class MergeSort {
    void mergesort(int a[], int n)
    {

        if (n < 2) {
            return;
        }
        int mid = n / 2;
        int[] l = new int[mid];
        int[] r = new int[n - mid];
    
        for (int i = 0; i < mid; i++) {
            l[i] = a[i];
        }
        for (int i = mid; i < n; i++) {
            r[i - mid] = a[i];
        }
        mergesort(l, mid);
        mergesort(r, n - mid);
    
        merge(a, l, r, mid, n - mid);

    }

    void merge(
  int[] a, int[] l, int[] r, int left, int right) {
 
    int i = 0, j = 0, k = 0;
    while (i < left && j < right) {
        if (l[i] <= r[j]) {
            a[k++] = l[i++];
        }
        else {
            a[k++] = r[j++];
        }
    }
    while (i < left) {
        a[k++] = l[i++];
    }
    while (j < right) {
        a[k++] = r[j++];
    }
}



  
    /* Prints the array */
    void printArray(int arr[])
    {
        int n = arr.length;
        for (int i = 0; i < n; ++i)
            System.out.print(arr[i] + " ");
        System.out.println();
    }
  
    // Driver method to test above
    public static void main(String args[])
    {
        MergeSort ob = new MergeSort();
        int arr[] = { 64, 34, 25, 12, 22, 11, 90 };
        ob.mergesort(arr, arr.length);
        System.out.println("Sorted array");
        ob.printArray(arr);
    }
}